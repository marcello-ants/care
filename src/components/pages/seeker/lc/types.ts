import { GeneralProviderProfile } from '@/types/seeker';

export interface ProviderProfile extends GeneralProviderProfile {
  numberOfReviews: number;
  qualities?: ProviderQualities;
  services?: ProviderServices;
  firstName: string;
}

export interface ProviderQualities {
  certifiedNursingAssistant?: boolean;
  certifiedRegisterNurse?: boolean;
  certifiedTeacher?: boolean;
  childDevelopmentAssociate?: boolean;
  comfortableWithPets?: boolean;
  cprTrained?: boolean;
  crn?: boolean;
  doesNotSmoke?: boolean;
  doula?: boolean;
  earlyChildDevelopmentCoursework?: boolean;
  earlyChildhoodEducation?: boolean;
  firstAidTraining?: boolean;
  nafccCertified?: boolean;
  ownTransportation?: boolean;
  specialNeedsCare?: boolean;
  trustlineCertifiedCalifornia?: boolean;
}

export interface ProviderServices {
  carpooling?: boolean;
  craftAssistance?: boolean;
  errands?: boolean;
  groceryShopping?: boolean;
  laundryAssistance?: boolean;
  lightHousekeeping?: boolean;
  mealPreparation?: boolean;
  swimmingSupervision?: boolean;
  travel?: boolean;
}
