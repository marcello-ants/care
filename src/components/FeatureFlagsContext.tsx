import { createContext, useContext, ReactNode, useState } from 'react';
import { LDEvaluationDetail } from 'launchdarkly-node-server-sdk';

interface FeatureFlagsProviderProps {
  children: ReactNode;
  flags: FeatureFlags;
}

export type featureFlagsState = {
  flags: FeatureFlags;
};

export type FeatureFlags = Record<string, LDEvaluationDetail | undefined>;

const initialState: featureFlagsState = {
  flags: {},
};

const FeatureFlagsContext = createContext<featureFlagsState>(initialState);

export const FeatureFlagsProvider = ({ children, flags }: FeatureFlagsProviderProps) => {
  const [state] = useState({ flags });
  return <FeatureFlagsContext.Provider value={state}>{children}</FeatureFlagsContext.Provider>;
};

export const useFeatureFlags = () => useContext(FeatureFlagsContext);
