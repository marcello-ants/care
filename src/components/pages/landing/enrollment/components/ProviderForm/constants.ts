export const PROVIDER_SERVICES_DESKTOP = [
  {
    description: 'Child care',
    label: 'Child care',
    value: 'CHILD_CARE',
  },
  {
    description: 'Senior care',
    label: 'Senior care',
    value: 'SENIOR_CARE',
  },
  {
    description: 'Housekeeping',
    label: 'Housekeeping',
    value: 'HOUSEKEEPING',
  },
  {
    description: 'Pet care',
    label: 'Pet care',
    value: 'PET_CARE',
  },
  {
    description: 'Tutoring',
    label: 'Tutoring',
    value: 'TUTORING',
  },
  {
    description: 'Special needs care',
    label: 'Special needs care',
    value: 'SPECIAL_NEEDS',
  },
  {
    description: 'Errands & odd jobs',
    label: 'Errands & odd jobs',
    value: 'GENERIC',
  },
];

export const PROVIDER_SERVICES_MOBILE = [
  {
    description: 'Child care',
    label: 'Child care',
    value: 'CHILD_CARE',
  },
  {
    description: 'Senior care',
    label: 'Senior care',
    value: 'SENIOR_CARE',
  },
  {
    description: 'Pet care',
    label: 'Pet care',
    value: 'PET_CARE',
  },
  {
    description: 'Tutoring',
    label: 'Tutoring',
    value: 'TUTORING',
  },
  {
    description: 'Special needs care',
    label: 'Special needs care',
    value: 'SPECIAL_NEEDS',
  },
  {
    description: 'Housekeeping',
    label: 'Housekeeping',
    value: 'HOUSEKEEPING',
  },
  {
    description: 'Errands & odd jobs',
    label: 'Errands & odd jobs',
    value: 'GENERIC',
  },
];

export const DROPOFF_URLS = {
  CHILD_CARE: '/swf/ccOptimizedFlow',
  SENIOR_CARE: '/app/enrollment/provider/sc ',
  PET_CARE: '/swf/pcOptimizedFlow',
  HOUSEKEEPING: '/swf/hkOptimizedFlow',
  TUTORING: '/swf/tuOptimizedFlow',
  SPECIAL_NEEDS: '/swf/snOptimizedFlow',
  GENERIC: '/swf/cgOptimizedFlow',
  AU_PAIR: '',
  CARE_GIGS: '',
};
