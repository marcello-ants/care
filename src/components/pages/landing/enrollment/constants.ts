export const SITTER_FORM = {
  providerServices: {
    title: 'What Services do you offer?',
    id: 'providerService',
    options: [
      {
        value: 'CHILD_CARE',
        text: 'Babysitter',
      },
      {
        value: 'CHILD_CARE',
        text: 'Nanny',
      },
      {
        value: 'SPECIAL_NEEDS',
        text: 'Special Needs Provider',
      },
      {
        value: 'TUTORING',
        text: 'Tutor',
      },
      {
        value: 'TUTORING',
        text: 'Private Lesson Instructor',
      },
      {
        value: 'SENIOR_CARE',
        text: 'Senior Care Provider',
      },
      {
        value: 'PET_CARE',
        text: 'Pet Care Provider',
      },
      {
        value: 'HOUSEKEEPING',
        text: 'Housekeeper',
      },
      {
        value: 'GENERIC',
        text: 'Errands &amp; Odd Jobs Provider',
      },
    ],
  },
  providerEmail: {
    id: 'email',
    placeholder: 'Email',
  },
  providerPassword: {
    id: 'password',
    placeholder: 'Password',
  },
  providerZip: {
    id: 'providerZip',
    placeholder: 'ZIP',
  },
};

export const GENERIC_FORM = {
  howDidYouHearAboutUs: {
    title: 'How did you hear about us?',
    id: 'howDidYouHearAboutUs',
    options: [
      {
        value: 'FACEBOOK',
        text: 'Facebook',
      },
      {
        value: 'BANNER_AD',
        text: 'Banner Ad',
      },
      {
        value: 'TV_AD',
        text: 'TV Ad',
      },
      {
        value: 'OTHER_SOCIAL_MEDIA',
        text: 'Other Social Media (Twitter, Instagram, Pinterest, LinkedIn)',
      },
      {
        value: 'PRESS_COVERAGE',
        text: 'Press Coverage (News, Magazine, Blog)',
      },
      {
        value: 'FRIENDS_OR_FAMILY',
        text: 'Friends or Family',
      },
      {
        value: 'ONLINE_VIDEO',
        text: 'Online Video (Hulu, Youtube)',
      },
      {
        value: 'SEARCH_ENGINE',
        text: 'Search Engine (Google, Bing)',
      },
      {
        value: 'PARENTING_GROUP',
        text: 'Parenting Group or Forum',
      },
      {
        value: 'OTHER',
        text: 'Other',
      },
    ],
  },
};
