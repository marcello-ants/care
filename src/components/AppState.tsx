import { useRouter } from 'next/router';
import { createContext, useReducer, useContext, ReactNode, useEffect } from 'react';
import { debounce } from 'lodash-es';
import { configureScope, captureException, Severity } from '@sentry/nextjs';
import {
  initialAppState,
  rootReducer,
  isStateVersionCompatible,
  doesStateHaveDefaultKeys,
  validateState,
} from '@/state';
import patchAppState from '@/state/patch';
import { initialState as flowInitialState } from '@/state/flow';
import logger from '@/lib/clientLogger';
import { ClientLogger } from '@care/utils';
import getVertical from '@/utilities/verticalUtils';
import { SessionStorageHelper } from '@/utilities/SessionStorageHelper';
import { trackPageViewInSiftScience } from '@/utilities/siftScienceHelper';
import { AppAction, AppDispatch, AppState } from '@/types/app';
import { StateCheckingResponse } from '@/types/common';
import {
  LOCAL_STORAGE_STATE_KEY,
  CZEN_BASE_PATH,
  FLOWS,
  POST_A_JOB_ROUTES,
  CLIENT_FEATURE_FLAGS,
} from '@/constants';
import { useFeatureFlags } from '@/components/FeatureFlagsContext';
import { AnalyticsHelper } from '@/utilities/analyticsHelper';

type AppStateProviderProps = {
  children: ReactNode;
  initialStateOverride?: AppState;
  czenJSessionId?: string | undefined;
  referrerCookie?: string | undefined | null;
  userHasAccount?: boolean;
  memberUuid?: string;
};

function getStateFromStorageOrDefault(defaultState: AppState) {
  let stateFromStorageJson;
  const sessionStorageAvailable = SessionStorageHelper.isAvailable();

  if (sessionStorageAvailable) {
    stateFromStorageJson = SessionStorageHelper.getItem(LOCAL_STORAGE_STATE_KEY);
  } else {
    logger.warn({ event: 'sessionStorageUnavailable' });
  }

  if (stateFromStorageJson) {
    const stateFromStorage = JSON.parse(stateFromStorageJson) as AppState;
    try {
      const patchedState = patchAppState(stateFromStorage, defaultState);

      if (isStateVersionCompatible(patchedState) && doesStateHaveDefaultKeys(patchedState)) {
        return patchedState;
      }
      logger.warn({ event: 'invalidStateFromLocalStorage', state: stateFromStorageJson });
    } catch (e) {
      captureException(e, { level: Severity.Warning });
    }
  }
  return defaultState;
}

function persistState(state: AppState) {
  SessionStorageHelper.setItem(LOCAL_STORAGE_STATE_KEY, JSON.stringify(state));
}

const debouncePersistState = debounce(persistState, 500, {
  leading: true,
  trailing: true,
  maxWait: 1000,
});

function reducerWrapper(state: AppState, action: AppAction) {
  const nextState = rootReducer(state, action);
  debouncePersistState(nextState);
  return nextState;
}

export function redirectIfInvalidState(state: AppState, pathname: string): StateCheckingResponse {
  if (
    typeof Intl === 'undefined' ||
    typeof Intl.NumberFormat === 'undefined' ||
    typeof Intl.DateTimeFormat === 'undefined'
  ) {
    // TODO: Remove this with SC-214
    logger.error({ event: 'IntlUnavailable', source: pathname, invalidState: state });
    return {
      redirectCalled: true,
      readableMessage: 'Safari 9 and lower is not currently supported',
    };
  }

  const isStateValid = validateState(state, pathname);
  const redirect = !isStateValid;

  if (redirect) {
    SessionStorageHelper.removeItem(LOCAL_STORAGE_STATE_KEY);

    window.location.assign(CZEN_BASE_PATH);
  }

  return { redirectCalled: redirect };
}

export function getFlowName(pathname: string, oldFlowName?: string): string {
  let newFlowName = Object.values(FLOWS).find((flow) => pathname.includes(flow.path))?.name;

  if (!newFlowName) {
    // TODO: this shouldn't be needed after SC-123 is implemented
    const isPostAJob = Object.values(POST_A_JOB_ROUTES).some((route) => pathname.includes(route));
    if (isPostAJob) {
      newFlowName = 'SEEKER_POST_A_JOB';
    }
  }

  // Don't allow shifting from in-facility to seeker flow name
  if (newFlowName === FLOWS.SEEKER.name && oldFlowName === FLOWS.SEEKER_IN_FACILITY.name) {
    return FLOWS.SEEKER_IN_FACILITY.name;
  }

  return newFlowName ?? '';
}

function useFlowDetection(dispatch: AppDispatch, oldFlowName: string) {
  const { pathname } = useRouter();
  const newFlowName = getFlowName(pathname, oldFlowName);
  const vertical = getVertical();
  ClientLogger.setLoggerStatics({ flow: newFlowName, vertical });

  useEffect(() => {
    dispatch({ type: 'setFlowName', flowName: newFlowName });

    configureScope((scope) => {
      scope.setTags({
        flow: newFlowName,
        vertical,
      });
    });
  }, [newFlowName]);
}

const AppStateContext = createContext<AppState>(initialAppState);
const AppDispatchContext = createContext<AppDispatch>(() =>
  // eslint-disable-next-line no-console
  console.warn(
    'Calling default dispatch is a no-op! You are probably consuming app dispatch in a component above AppStateProvider'
  )
);

export const AppStateProvider = ({
  children,
  initialStateOverride,
  czenJSessionId,
  referrerCookie,
  userHasAccount,
  memberUuid,
}: AppStateProviderProps) => {
  const hydratedInitialAppState = {
    ...initialAppState,
    flow: {
      ...flowInitialState,
      czenJSessionId,
      referrerCookie,
      userHasAccount,
      memberUuid,
    },
  };
  const featureFlags = useFeatureFlags();
  const sessionStorageRetrievalSuccessEvaluation =
    featureFlags?.flags[CLIENT_FEATURE_FLAGS.SESSION_STORAGE_RETRIEVAL_SUCCESS];
  const shouldMeasureSessionStorageRetrievalSuccess =
    sessionStorageRetrievalSuccessEvaluation?.variationIndex === 1;

  const router = useRouter();

  const logSessionStorageError = () => {
    logger.info({ event: 'sessionStorageStateUnavailable', path: router.pathname });
    AnalyticsHelper.logEvent({
      name: 'sessionStorageStateUnavailable',
    });
  };
  const launchDarklyWrappedGetStateFromStorageOrDefault = (defaultState: AppState) => {
    if (shouldMeasureSessionStorageRetrievalSuccess) {
      try {
        const sessionStorageState = SessionStorageHelper.getItem(LOCAL_STORAGE_STATE_KEY);
        if (sessionStorageState && sessionStorageState.length > 0) {
          const stateFromStorage = JSON.parse(sessionStorageState);
          if (stateFromStorage) {
            logger.info({ event: 'sessionStorageStateAvailable', path: router.pathname });
          } else {
            logSessionStorageError();
          }
        } else {
          logSessionStorageError();
        }
      } catch (_e) {
        logSessionStorageError();
      }
    }
    return getStateFromStorageOrDefault(defaultState);
  };

  const [state, dispatch] = useReducer(
    reducerWrapper,
    initialStateOverride ?? hydratedInitialAppState,
    launchDarklyWrappedGetStateFromStorageOrDefault
  );
  useFlowDetection(dispatch, state.flow.flowName);
  const routePath = router.asPath?.split('?')[0];
  const stateCheckResult = redirectIfInvalidState(state, routePath);
  if (stateCheckResult.redirectCalled) {
    if (stateCheckResult.readableMessage) {
      return (
        <>
          <p>Redirecting to Member Home Page...</p>
          <p>{stateCheckResult.readableMessage}</p>
        </>
      );
    }
    return <></>;
  }

  useEffect(() => {
    trackPageViewInSiftScience(state);
  }, [routePath]);

  return (
    <AppDispatchContext.Provider value={dispatch}>
      <AppStateContext.Provider value={state}>{children}</AppStateContext.Provider>
    </AppDispatchContext.Provider>
  );
};

AppStateProvider.defaultProps = {
  initialStateOverride: undefined,
  czenJSessionId: undefined,
  referrerCookie: undefined,
  userHasAccount: false,
  memberUuid: undefined,
};

export const useAppState = () => useContext(AppStateContext);
export const useAppDispatch = () => useContext(AppDispatchContext);

export const useSeekerState = () => useContext(AppStateContext).seeker;
export const useProviderState = () => useContext(AppStateContext).provider;
export const useProviderCCState = () => useContext(AppStateContext).providerCC;
export const useFlowState = () => useContext(AppStateContext).flow;
export const useSeekerCCState = () => useContext(AppStateContext).seekerCC;
export const useSeekerHKState = () => useContext(AppStateContext).seekerHK;
export const useSeekerPCState = () => useContext(AppStateContext).seekerPC;
export const useSeekerTUState = () => useContext(AppStateContext).seekerTU;
export const useLtcgState = () => useContext(AppStateContext).ltcg;
