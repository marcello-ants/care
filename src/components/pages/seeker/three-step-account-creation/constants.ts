import * as yup from 'yup';
import { ServiceType } from '@/__generated__/globalTypes';
import { VerticalsAbbreviation } from '@/constants';

export const validationSchema = yup.object({
  firstName: yup
    .string()
    .min(2, 'First name needs to have at least two letters.')
    .required('First name is required'),
  lastName: yup
    .string()
    .min(2, 'Last name needs to have at least two letters.')
    .required('Last name is required'),
});

export const SERVICE_TYPE_BY_VERTICAL: { [key: string]: ServiceType } = {
  HK: ServiceType.HOUSEKEEPING,
  PC: ServiceType.PET_CARE,
  TU: ServiceType.TUTORING,
  CC: ServiceType.CHILD_CARE,
  DC: ServiceType.CHILD_CARE,
  IB: ServiceType.CHILD_CARE,
  SC: ServiceType.SENIOR_CARE,
};

export const ERROR_CREATING_ACCOUNT = 'An error occurred creating your account, please try again.';

export const FACEBOOK_VERTICALS_ALLOWANCE = ['CC'];
export const HDYHAU_VERTICALS_ALLOWANCE: VerticalsAbbreviation[] = ['HK', 'TU', 'PC', 'DC'];
