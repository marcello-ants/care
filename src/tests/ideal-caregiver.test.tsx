import React from 'react';
import { render, waitFor, screen } from '@testing-library/react';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { NextRouter, useRouter } from 'next/router';
import { initialAppState } from '@/state';
import { AppStateProvider, useAppDispatch } from '@/components/AppState';
import { AppState } from '@/types/app';

import IdealCaregiver from '@/pages/ideal-caregiver';
import { SENIOR_CARE_RECURRING_JOB_CREATE } from '@/components/request/GQL';
import { CLIENT_FEATURE_FLAGS, POST_A_JOB_ROUTES } from '@/constants';
import { ServerError } from '@apollo/client/core';
import { generateJobCreateInput } from '@/utilities/gqlPayloadHelper';
import logger from '@/lib/clientLogger';
import { SENIOR_CARE_TYPE } from '@/__generated__/globalTypes';
import { useFeatureFlags } from '@/components/FeatureFlagsContext';
import AmpliHelper from '@/utilities/ampliAnalyticsHelper';
import { AnalyticsHelper } from '@/utilities/analyticsHelper';

jest.mock('@/components/FeatureFlagsContext', () => ({
  useFeatureFlags: jest.fn().mockReturnValue({ flags: {} }),
}));

jest.mock('next/router', () => ({
  __esModule: true,
  useRouter: jest.fn(),
}));

jest.mock('next/config', () => () => {
  return {
    publicRuntimeConfig: {
      OIDC_STS_AUTHORITY: 'OIDC_STS_AUTHORITY',
      OIDC_CLIENT_ID: 'OIDC_CLIENT_ID',
      OIDC_RESPONSE_TYPE: 'OIDC_RESPONSE_TYPE',
      OIDC_CLIENT_SCOPE: 'OIDC_CLIENT_SCOPE',
      OIDC_CALLBACK_PATH: 'OIDC_CALLBACK_PATH',
      OIDC_LOGOUT_URL: 'OIDC_LOGOUT_URL',
    },
  };
});

let mockRouter: Pick<NextRouter, 'push' | 'pathname'>;

jest.mock('@/components/AppState', () => ({
  ...(jest.requireActual('@/components/AppState') as object),
  useAppDispatch: jest.fn(),
}));
let mockAppDispatch: ReturnType<typeof useAppDispatch>;

const initialState: AppState = {
  ...initialAppState,
  seeker: {
    ...initialAppState.seeker,
    typeOfCare: SENIOR_CARE_TYPE.IN_HOME,
  },
};

const postAJobInput = generateJobCreateInput(initialState.seeker.jobPost);

const postAJobMock: MockedResponse = {
  request: {
    query: SENIOR_CARE_RECURRING_JOB_CREATE,
    variables: {
      input: postAJobInput,
    },
  },
  result: {
    data: {
      seniorCareRecurringJobCreate: {
        job: { id: 123 },
        __typename: 'SeniorCareRecurringJobCreatePayload',
      },
    },
  },
};

const inappropiateLanguageDescription =
  'This is a description with InappropiateLanguage, there are some forbidden words like bath or naked, this is an example. This is a description with InappropiateLanguage, there are some forbidden words like bath or naked, this is an example.';

const innapropiateLanguageState = {
  ...initialState,
  seeker: {
    ...initialState.seeker,
    jobPost: {
      ...initialState.seeker.jobPost,
      idealCaregiver: {
        ...initialState.seeker.jobPost.idealCaregiver,
        details: inappropiateLanguageDescription,
      },
    },
  },
};
const postAJobInappropiateLanguageInput = generateJobCreateInput(
  innapropiateLanguageState.seeker.jobPost
);
const postAJobInappropiateLanguageMock: MockedResponse = {
  request: {
    query: SENIOR_CARE_RECURRING_JOB_CREATE,
    variables: {
      input: {
        ...postAJobInappropiateLanguageInput,
      },
    },
  },
  result: {
    data: {
      seniorCareRecurringJobCreate: {
        errors: [
          {
            message:
              'Inappropriate language detected. Please check the information you entered in the Job Description field.',
            __typename: 'InappropriateLanguageInput',
          },
        ],
        __typename: 'SeniorCareRecurringJobCreateError',
      },
    },
  },
};

const authError: ServerError = {
  statusCode: 401,
} as any;
const postAJobAuthFailureMock: MockedResponse = {
  ...postAJobMock,
  error: authError,
};

function renderPage(state: AppState, mocks: MockedResponse[] = []) {
  return render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <AppStateProvider initialStateOverride={state}>
        <IdealCaregiver />
      </AppStateProvider>
    </MockedProvider>
  );
}

describe('Seeker - Senior Care - Ideal Caregiver', () => {
  const realLoggerWarn = logger.warn;
  const realLoggerError = logger.error;

  const loggerWarnMock = jest.fn();
  const loggerErrorMock = jest.fn();

  beforeAll(() => {
    logger.warn = loggerWarnMock;
    logger.error = loggerErrorMock;
  });

  afterAll(() => {
    logger.warn = realLoggerWarn;
    logger.error = realLoggerError;
  });
  beforeEach(() => {
    mockAppDispatch = jest.fn();
    (useAppDispatch as jest.Mock).mockReturnValue(mockAppDispatch);

    mockRouter = {
      push: jest.fn(), // the component uses `router.push` only
      pathname: '/ideal-caregiver',
    };
    (useRouter as jest.Mock).mockReturnValue(mockRouter);

    loggerWarnMock.mockReset();
    loggerErrorMock.mockReset();
  });

  it('matches snapshot', async () => {
    const { asFragment } = renderPage(initialState);
    const submitButton = screen.getByRole('button', { name: 'Submit' });
    await waitFor(() => expect(submitButton).toBeEnabled());

    expect(asFragment()).toMatchSnapshot();
  });

  it('should redirect the user to the next page after Submit click results in a successful job post', async () => {
    renderPage(initialState, [postAJobMock]);
    const submitButton = screen.getByRole('button', { name: 'Submit' });
    await waitFor(() => expect(submitButton).toBeEnabled());

    submitButton.click();

    await waitFor(() =>
      // should have dispatched than a submission was attempted
      expect(mockAppDispatch).toHaveBeenCalledWith({
        type: 'setSubmissionAttempted',
        submissionAttempted: true,
        attempts: 1,
      })
    );

    await waitFor(() =>
      // should have dispatched another message clearing that data after a successful PAJ
      expect(mockAppDispatch).toHaveBeenCalledWith({
        type: 'setSubmissionAttempted',
        submissionAttempted: false,
        attempts: 0,
      })
    );

    await waitFor(() =>
      expect(mockAppDispatch).toHaveBeenCalledWith({ type: 'setJobStatus', jobSuccessful: true })
    );
    await waitFor(() =>
      expect(mockRouter.push).toHaveBeenCalledWith(POST_A_JOB_ROUTES.CAREGIVERS_NEAR_YOU)
    );
  });

  it('should log a warning after Submit click results in an auth error', async () => {
    renderPage(initialState, [postAJobAuthFailureMock]);
    const submitButton = screen.getByRole('button', { name: 'Submit' });
    await waitFor(() => expect(submitButton).toBeEnabled());

    submitButton.click();

    await waitFor(() =>
      expect(mockAppDispatch).toHaveBeenCalledWith({
        type: 'setSubmissionAttempted',
        submissionAttempted: true,
        attempts: 1,
      })
    );
    await waitFor(() => expect(loggerErrorMock).not.toHaveBeenCalled());
    await waitFor(() => expect(loggerWarnMock).toHaveBeenCalledTimes(1));
    await waitFor(() =>
      expect(loggerWarnMock).toHaveBeenCalledWith({
        event: 'postAJobErrorFrom401',
        graphQLError: 'undefined',
      })
    );
  });

  it('should execute PAJ request on page load if an attempt was previously made by the user', async () => {
    const state = {
      ...initialState,
      seeker: {
        ...initialState.seeker,
        jobPost: {
          ...initialState.seeker.jobPost,
          idealCaregiver: {
            ...initialState.seeker.jobPost.idealCaregiver,
            submissionAttempted: true,
            attempts: 1,
          },
        },
      },
    };

    renderPage(state, [postAJobMock]);

    await waitFor(() =>
      expect(mockAppDispatch).toHaveBeenCalledWith({
        type: 'setSubmissionAttempted',
        submissionAttempted: false,
        attempts: 0,
      })
    );
    await waitFor(() =>
      expect(mockAppDispatch).toHaveBeenCalledWith({ type: 'setJobStatus', jobSuccessful: true })
    );
    await waitFor(() =>
      expect(mockRouter.push).toHaveBeenCalledWith(POST_A_JOB_ROUTES.CAREGIVERS_NEAR_YOU)
    );
  });

  it('should execute PAJ request on page load and log a warning if another auth error is received', async () => {
    const state = {
      ...initialState,
      seeker: {
        ...initialState.seeker,
        jobPost: {
          ...initialState.seeker.jobPost,
          idealCaregiver: {
            ...initialState.seeker.jobPost.idealCaregiver,
            submissionAttempted: true,
            attempts: 1,
          },
        },
      },
    };

    renderPage(state, [postAJobAuthFailureMock]);

    await waitFor(() =>
      expect(mockAppDispatch).toHaveBeenCalledWith({
        type: 'setSubmissionAttempted',
        submissionAttempted: true,
        attempts: 2,
      })
    );

    await waitFor(() => expect(loggerErrorMock).not.toHaveBeenCalled());
    await waitFor(() => expect(loggerWarnMock).toHaveBeenCalledTimes(1));
    await waitFor(() =>
      expect(loggerWarnMock).toHaveBeenCalledWith({
        event: 'postAJobErrorFrom401',
        graphQLError: 'undefined',
      })
    );
  });

  it('should execute PAJ request on page load and log an error when user is at max number of retries and if another auth error is received', async () => {
    const state = {
      ...initialState,
      seeker: {
        ...initialState.seeker,
        jobPost: {
          ...initialState.seeker.jobPost,
          idealCaregiver: {
            ...initialState.seeker.jobPost.idealCaregiver,
            submissionAttempted: true,
            attempts: 3,
          },
        },
      },
    };

    renderPage(state, [postAJobAuthFailureMock]);

    await waitFor(() => expect(loggerWarnMock).not.toHaveBeenCalled());
    await waitFor(() =>
      expect(loggerErrorMock).toHaveBeenCalledWith({
        event: 'postAJobMaxAttemptsFrom401',
        graphQLError: 'undefined',
      })
    );
    await waitFor(() =>
      expect(mockRouter.push).not.toHaveBeenCalledWith(POST_A_JOB_ROUTES.CAREGIVERS_NEAR_YOU)
    );
  });

  it('should show an error if used innapropiate language description', async () => {
    renderPage(innapropiateLanguageState, [postAJobInappropiateLanguageMock]);
    const submitButton = screen.getByRole('button', { name: 'Submit' });
    await waitFor(() => expect(submitButton).toBeEnabled());

    submitButton.click();

    await waitFor(() =>
      expect(loggerErrorMock).toHaveBeenCalledWith({
        event: 'postAJobInappropriateLanguageInput',
        graphQLError:
          'Inappropriate language detected. Please check the information you entered in the Job Description field.',
      })
    );
  });

  it('should fire both analytic events when feature flag is on', async () => {
    const ampliListener = jest.spyOn(AmpliHelper.ampli, 'jobPostedCaregiverNeeds');
    const amplitudeListener = jest.spyOn(AnalyticsHelper, 'logEvent');

    (useFeatureFlags as jest.Mock).mockReturnValue({
      flags: { [CLIENT_FEATURE_FLAGS.AMPLITUDE_USE_AMPLI]: { variationIndex: 1 } },
    });

    renderPage(initialState, [postAJobMock]);

    const submitButton = screen.getByRole('button', { name: 'Submit' });
    await waitFor(() => expect(submitButton).toBeEnabled());
    submitButton.click();

    await waitFor(() => {
      expect(ampliListener).toBeCalledTimes(1);
    });

    await waitFor(() => {
      expect(amplitudeListener).toBeCalledTimes(1);
    });
  });
});
