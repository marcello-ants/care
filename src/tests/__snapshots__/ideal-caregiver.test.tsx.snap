// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Seeker - Senior Care - Ideal Caregiver matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h2
        class="MuiTypography-root makeStyles-header-13 MuiTypography-h2"
      >
        What are you looking for in a caregiver?
      </h2>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h4
        class="MuiTypography-root makeStyles-idealCaregiverSubtitle-8 MuiTypography-h4"
      >
        Is there anything else we should know about your ideal caregiver?
      </h4>
    </div>
    <div
      class="MuiGrid-root makeStyles-caregiverContainer-2 MuiGrid-item MuiGrid-grid-xs-12"
    >
      <p
        class="MuiTypography-root makeStyles-examplesText-9 MuiTypography-body2"
      >
        Examples
      </p>
      <ul
        class="MuiList-root makeStyles-bulletPointList-4 MuiList-padding"
      >
        <li
          class="MuiListItem-root makeStyles-bulletPointItem-5"
        >
          <p
            class="MuiTypography-root makeStyles-body3-17 MuiTypography-body1"
          >
            What type of personality would be a good fit?
          </p>
        </li>
        <li
          class="MuiListItem-root makeStyles-bulletPointItem-5"
        >
          <p
            class="MuiTypography-root makeStyles-body3-17 MuiTypography-body1"
          >
            Are there any specialized skills they need?
          </p>
        </li>
        <li
          class="MuiListItem-root makeStyles-bulletPointItem-5"
        >
          <p
            class="MuiTypography-root makeStyles-body3-17 MuiTypography-body1"
          >
            Should they have any specific hobbies/interests?
          </p>
        </li>
      </ul>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <textarea
        class="makeStyles-detailsBox-6"
        id="details"
        name="details"
        placeholder="Share details here"
        rows="7"
        style="height: -4px; overflow: hidden;"
      />
      <textarea
        aria-hidden="true"
        class="makeStyles-detailsBox-6"
        readonly=""
        style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
        tabindex="-1"
      />
    </div>
    <div
      class="MuiGrid-root makeStyles-buttonsContainer-12 MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="makeStyles-submitButtonContainer-3"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Submit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
